Add CLI option to choose model checker engine

leonardoalt opened the pull request:
Fixes https://github.com/ethereum/solidity/issues/9899

Commit:
Add CLI option to choose model checker engine

leonardoalt commented on solc/CommandLineInterface.h:
```diff
@@ -79,7 +79,6 @@ class CommandLineInterface
 	void handleABI(std::string const& _contract);
 	void handleNatspec(bool _natspecDev, std::string const& _contract);
 	void handleGasEstimation(std::string const& _contract);
-	void handleFormal();
```
This was deprecated long time ago.

cameel replied:
Yeah, looks like there's no definition and no references to it anywhere.

cameel commented on libsolidity/formal/ModelChecker.h:
```diff
@@ -41,6 +41,21 @@ struct SourceLocation;
+	static constexpr ModelCheckerEngine None() { return {false, false}; }
+
+	bool none() { return !some(); }
+	bool some() { return bmc || chc; }
```
How about `any()`?
```suggestion
	bool any() { return bmc || chc; }
```

cameel commented on libsolidity/formal/ModelChecker.h:
```diff
@@ -41,6 +41,21 @@ struct SourceLocation;
+	static constexpr ModelCheckerEngine CHC() { return {false, true}; }
+	static constexpr ModelCheckerEngine None() { return {false, false}; }
+
+	bool none() { return !some(); }
+	bool some() { return bmc || chc; }
+	bool all() { return bmc && chc; }
```
```suggestion
	bool none() const { return !some(); }
	bool some() const { return bmc || chc; }
	bool all() const { return bmc && chc; }
```

cameel commented on solc/CommandLineInterface.cpp:
```diff
@@ -1379,6 +1391,24 @@ bool CommandLineInterface::processInput()
+	if (m_args.count(g_argModelCheckerEngine))
+	{
+		string engine = m_args[g_argModelCheckerEngine].as<string>();
+		if (engine == "none")
+			m_modelCheckerEngine = ModelCheckerEngine::None();
+		else if (engine == "bmc")
+			m_modelCheckerEngine = ModelCheckerEngine::BMC();
+		else if (engine == "chc")
+			m_modelCheckerEngine = ModelCheckerEngine::CHC();
+		else if (engine == "all")
+			m_modelCheckerEngine = ModelCheckerEngine::All();
```
I think these strings should be globally defined in something like `g_modelCheckerEngineArgs`. That's how it's done for other options with arguments here.

leonardoalt replied:
Not all, some are also done via strings. I can do that too. I did it this way to avoid more of these small definitions in this file.

cameel replied:
Ah, ok. These globals do add some clutter so if you want to avoid that it's fine with me. Personally I'd use them simply because with strings it's too easy to make a typo that won't be detected.

leonardoalt replied:
Yea I agree with you that the globals are safer, will change.

cameel replied:
I see that you defined them but you're still not using them here.

leonardoalt replied:
True, forgot. Will update

leonardoalt replied:
Now I'm actually only using them to display the options, since I'm using function `ModelChecker::fromString` for the conversion. Not sure it makes sense to keep the variables just for that @cameel

cameel replied:
I mean, if you already have the variables I'd use them. If you want to remove them that's also fine with me as I said above.

cameel commented on solc/CommandLineInterface.cpp:
```diff
@@ -993,6 +995,16 @@ General Information)").c_str(),
+	smtCheckerOptions.add_options()
+		(
+			g_strModelCheckerEngine.c_str(),
+			po::value<string>()->value_name("none, bmc, chc, all"),
```
If you define `g_modelCheckerEngineArgs`, this could be `boost::join(g_modelCheckerEngineArgs, ",")`.

cameel commented on solc/CommandLineInterface.cpp:
```diff
@@ -993,6 +995,16 @@ General Information)").c_str(),
+	smtCheckerOptions.add_options()
+		(
+			g_strModelCheckerEngine.c_str(),
+			po::value<string>()->value_name("none, bmc, chc, all"),
```
```suggestion
			po::value<string>()->value_name("none, bmc, chc, all")->default_value("all"),
```
This way the default will be displayed in the command-line help.

Also, with that added here the default on `m_modelCheckerEngine` in class definition seems superfluous, though I'm not sure if we could remove it easily because the struct needs to be initialized...

cameel replied:
Looks like you forgot about `default_value()` :)

leonardoalt replied:
Fixed

ekpyron commented:
What about standard json?

cameel commented:
Right. Standard JSON. And then it need to be added to the docs for Standard JSON input.

Also, having two command-line tests to make sure it actually works (one for standard json and one for normal command-line) might be a good idea.

ekpyron commented:
I was mainly wondering, though - standard json can be a separate PR.

cameel commented:
Could be but I found it frustrating at times that some options are only in one or the other. Whether here or in a separate PR, I think it's a good idea to have all newly added options in both.

ekpyron commented:
Yeah, the other PR should definitely go into the same release :-).
By the way: in general standard-json is more feature rich, so usually that'd be the first place to implement it and then have the CLI follow (if both is not done at once).
But yeah, I'm not sure what @leonardoalt's plan was for that here :-).

leonardoalt commented:
I was hoping no one would remember standard json and we could go without :p
I planned on adding it later, mostly because the people that need this feature right now use the compiler via CLI. But I can also do it in this PR.

cameel commented:
Reasons for #9583 keep piling up :)

leonardoalt commented:
- Added model checker engine strings instead of hardcoded strings
- Added standard json option
- Added tests

Ready for review.

@ekpyron Actually after this we can have CHC only counterexample tests as well

stackenbotten commented:
There was an error when running `chk_coding_style` for commit `17325f71fe48cab2ac370247f8a2af5bea300100`:
```
Error: Trailing whitespace found:
 test/cmdlineTests/model_checker_engine_all/err:6:Note: 
 test/cmdlineTests/model_checker_engine_all/err:14:State: 
 test/cmdlineTests/model_checker_engine_bmc/err:6:Note: 
 test/cmdlineTests/model_checker_engine_bmc/err:10:Note: Callstack: 
 test/cmdlineTests/model_checker_engine_bmc/err:11:Note: 
 test/cmdlineTests/model_checker_engine_chc/err:6:Note: 
 test/cmdlineTests/model_checker_engine_chc/err:14:State: 
 test/cmdlineTests/standard_model_checker_engine_all/output.json:12:State: 
 test/cmdlineTests/standard_model_checker_engine_all/output.json:22:State: 
 test/cmdlineTests/standard_model_checker_engine_bmc/output.json:27:Callstack: 
 test/cmdlineTests/standard_model_checker_engine_chc/output.json:12:State: 
 test/cmdlineTests/standard_model_checker_engine_chc/output.json:22:State: 

```
Please check that your changes are working as intended.

leonardoalt commented:
> There was an error when running `chk_coding_style` for commit `17325f71fe48cab2ac370247f8a2af5bea300100`:
> 
> ```
> Error: Trailing whitespace found:
>  test/cmdlineTests/model_checker_engine_all/err:6:Note: 
>  test/cmdlineTests/model_checker_engine_all/err:14:State: 
>  test/cmdlineTests/model_checker_engine_bmc/err:6:Note: 
>  test/cmdlineTests/model_checker_engine_bmc/err:10:Note: Callstack: 
>  test/cmdlineTests/model_checker_engine_bmc/err:11:Note: 
>  test/cmdlineTests/model_checker_engine_chc/err:6:Note: 
>  test/cmdlineTests/model_checker_engine_chc/err:14:State: 
>  test/cmdlineTests/standard_model_checker_engine_all/output.json:12:State: 
>  test/cmdlineTests/standard_model_checker_engine_all/output.json:22:State: 
>  test/cmdlineTests/standard_model_checker_engine_bmc/output.json:27:Callstack: 
>  test/cmdlineTests/standard_model_checker_engine_chc/output.json:12:State: 
>  test/cmdlineTests/standard_model_checker_engine_chc/output.json:22:State: 
> ```
> 
> Please check that your changes are working as intended.

This is annoying though. Not sure I can get rid of those trailing whitespaces that easily

ekpyron commented:
> > There was an error when running `chk_coding_style` for commit `17325f71fe48cab2ac370247f8a2af5bea300100`:
> > ```
> > Error: Trailing whitespace found:
> >  test/cmdlineTests/model_checker_engine_all/err:6:Note: 
> >  test/cmdlineTests/model_checker_engine_all/err:14:State: 
> >  test/cmdlineTests/model_checker_engine_bmc/err:6:Note: 
> >  test/cmdlineTests/model_checker_engine_bmc/err:10:Note: Callstack: 
> >  test/cmdlineTests/model_checker_engine_bmc/err:11:Note: 
> >  test/cmdlineTests/model_checker_engine_chc/err:6:Note: 
> >  test/cmdlineTests/model_checker_engine_chc/err:14:State: 
> >  test/cmdlineTests/standard_model_checker_engine_all/output.json:12:State: 
> >  test/cmdlineTests/standard_model_checker_engine_all/output.json:22:State: 
> >  test/cmdlineTests/standard_model_checker_engine_bmc/output.json:27:Callstack: 
> >  test/cmdlineTests/standard_model_checker_engine_chc/output.json:12:State: 
> >  test/cmdlineTests/standard_model_checker_engine_chc/output.json:22:State: 
> > ```
> > 
> > 
> > Please check that your changes are working as intended.
> 
> This is annoying though. Not sure I can get rid of those trailing whitespaces that easily

Looks pretty easy to me...
CHC.cpp:L1407 for example - just don't output the space if ``formatVariableModel(*stateVars, stateValues, ", ")`` is empty or starts with a linebreak.

leonardoalt commented:
That one is easy. But for example
```
Note:<whitespace>
Counterexample:
```
I don't want `Counterexample` to be in front of `Note:`, but `Note:` always adds a whitespace anyway, so the only solution here, without changing the internals too much, is to indeed add `Note: Counterexample:` which to me looks ugly

cameel commented on test/cmdlineTests/model_checker_engine_all/exit:
```diff
@@ -0,0 +1 @@
+0
```
I think that `exit` file is optional if the code is zero so you might remove it to have less clutter...

cameel replied:
`err` is optional too if there's no output on stderr.

cameel commented on test/cmdlineTests/standard_model_checker_engine_all/input.json:
```diff
@@ -0,0 +1,14 @@
+			"content": "// SPDX-License-Identifier: GPL-3.0\npragma solidity >=0.0;\npragma experimental SMTChecker;\ncontract C { function f(uint x) public pure { assert(x > 0); } }"
+		}
+	},
+	"settings":
+	{
+		"modelCheckerEngine": "all"
+	}
```
Settings key is completely missing from the output description in the docs. I think it should be added there, along with your new option. Or is it omitted because it just repeats all the input settings?

cameel replied:
Oops. This is an input file. Sorry, disregard this comment :)

cameel commented on docs/using-the-compiler.rst:
```diff
@@ -341,7 +341,9 @@ Input Description
-        }
+        },
+        // Choose which model checker engine to use: all (default), bmc, chc, none.
+        "modelCheckerEngine": "chc"
```
Do we want to include the state of this setting in medatata (in `CompilerStack::createMetadata()`)? Or is it not necessary because SMT checker cannot affect compilation output?

leonardoalt replied:
Yea I think it's not necessarily because it doesn't affect bytecode generation.

cameel replied:
OK. Fine.

ekpyron commented:
> That one is easy. But for example
> 
> ```
> Note:<whitespace>
> Counterexample:
> ```
> 
> I don't want `Counterexample` to be in front of `Note:`, but `Note:` always adds a whitespace anyway, so the only solution here, without changing the internals too much, is to indeed add `Note: Counterexample:` which to me looks ugly

Just change liblangutil/SourceReferenceFormatterHuman.cpp:L168 and have it only add the space if ``secondary.message`` is nonempty...

leonardoalt commented:
I changed it to output `Note: Counterexample:`

ekpyron commented:
Or that.

leonardoalt commented on solc/CommandLineInterface.cpp:
```diff
@@ -1379,6 +1403,18 @@ bool CommandLineInterface::processInput()
+	if (m_args.count(g_argModelCheckerEngine))
+	{
+		string engineStr = m_args[g_argModelCheckerEngine].as<string>();
+		optional<ModelCheckerEngine> engine = ModelCheckerEngine::fromString(engineStr);
```
@cameel This is what I mean. If I use the variables here this block will actually get larger.

cameel replied:
Oh, thanks for the pointer. I was actually looking at the old version.

In that case obviously I wouldn't use them here.

For `value_name()` it would be best if you just moved the variables to SMT checker and used the same constants in both it and in `fromString()`... but to be honest it's additional hassle for a marginal gain so at this point I'd just leave this as it is now.

cameel replied:
Or, if you want to remove them, I'm fine with that too. For me the biggest benefit is in avoiding accidental typos. With how short and few they are, that's not a big risk here.

leonardoalt replied:
Yea, I think it's overkill to add variables for them.

leonardoalt commented:
Does anyone have any idea why the test fails?
It says
```
- model_checker_engine_none/
Incorrect output on stderr received. Expected:

But got:
Compiler run successful, no output requested.
When running /root/project/build/solc/solc model_checker_engine_none/input.sol --model-checker-engine none </dev/stdin
```
But my local run indeed outputs nothing.

cameel commented:
Maybe it outputs whitespace?

leonardoalt commented:
Maybe I need the `stderr` file to exist so it actually gets updated by the tool.

cameel commented on libsolidity/formal/BMC.cpp:
```diff
@@ -863,13 +864,18 @@ void BMC::checkCondition(
 			for (auto const& eval: sortedModel)
 				modelMessage << "  " << eval.first << " = " << eval.second << "\n";
+
+			auto extraInfo = SecondarySourceLocation().append(modelMessage.str(), SourceLocation{});
+			auto callStack = SMTEncoder::callStackMessage(_callStack);
+			if (callStack.infos.size() > 0)
+				extraInfo = extraInfo.append(SMTEncoder::callStackMessage(_callStack));
+			if (secondaryLocation.infos.size() > 0)
+				extraInfo = extraInfo.append(move(secondaryLocation));
```
Maybe it would be better to add the non-empty check in `SecondarySourceLocation.append()` to solve the issue more generally? Or add another method like `appendIfNotEmpty()`?

cameel replied:
Actually, I just noticed in a different place:
```c++
		if (call.second)
			callStackLocation.append("", call.second->location())
```
So apparently we sometimes do want the string to be empty?

leonardoalt replied:
Yea, in this case I just want to print the SourceLocation. The two ifs in BMC.cpp were actually not necessary because `SecondaryLocation::append(SecondaryLocation&& _other)` simply does `infos += other.infos`, so if it's empty nothing happens anyway.

leonardoalt commented:
No, still... my local test still tells me the files are fine

cameel commented:
Apparently the compiler prints a message to stderr when there's no output :)

```c++
	if (!g_hasOutput)
	{
		if (m_args.count(g_argOutputDir))
			sout() << "Compiler run successful. Artifact(s) can be found in directory " << m_args.at(g_argOutputDir).as<string>() << "." << endl;
		else
			serr() << "Compiler run successful, no output requested." << endl;
	}
```

leonardoalt commented:
> Apparently the compiler prints a message to stderr when there's no output :)
> 
> ```c++
> 	if (!g_hasOutput)
> 	{
> 		if (m_args.count(g_argOutputDir))
> 			sout() << "Compiler run successful. Artifact(s) can be found in directory " << m_args.at(g_argOutputDir).as<string>() << "." << endl;
> 		else
> 			serr() << "Compiler run successful, no output requested." << endl;
> 	}
> ```

I wonder why the script is not complaining locally though

cameel commented on libsolidity/formal/SMTEncoder.cpp:
```diff
@@ -2114,10 +2114,12 @@ SecondarySourceLocation SMTEncoder::callStackMessage(vector<CallStackEntry> cons
 		if (call.second)
 			callStackLocation.append("", call.second->location());
+	if (callStackLocation.infos.size() == 1)
+		return {};
```
Why this change?

I don't see the callstack printed in any of the tests so I can't tell if it's right or wrong...

leonardoalt replied:
I changed how the secondarysource location is printed, so maybe this won't be necessary

cameel commented:
For me it outputs this locally:

```json
{"errors":[{"component":"general","errorCode":"3805","formattedMessage":"Warning: This is a pre-release compiler version, please do not use it in production.\n","message":"This is a pre-release compiler version, please do not use it in production.","severity":"warning","type":"Warning"}],"sources":{"A":{"id":0}}}
```

But if the pre-release warning is the problem here then other tests should be failing too...

cameel commented:
Oh, I see that the pre-release warning gets stripped by the test script. The stderr output does not.

```bash
    if [[ "$solc_args" == *"--standard-json"* ]]
    then
        sed -i.bak -e 's/{[^{]*Warning: This is a pre-release compiler version[^}]*},\{0,1\}//' "$stdout_path"
        sed -i.bak -E -e 's/ Consider adding \\"pragma solidity \^[0-9.]*;\\"//g' "$stdout_path"
        sed -i.bak -e 's/"errors":\[\],\{0,1\}//' "$stdout_path"
        # Remove explicit bytecode and references to bytecode offsets
        sed -i.bak -E -e 's/\"object\":\"[a-f0-9]+\"/\"object\":\"bytecode removed\"/g' "$stdout_path"
        sed -i.bak -E -e 's/\"opcodes\":\"[^"]+\"/\"opcodes\":\"opcodes removed\"/g' "$stdout_path"
        sed -i.bak -E -e 's/\"sourceMap\":\"[0-9:;-]+\"/\"sourceMap\":\"sourceMap removed\"/g' "$stdout_path"
        # Replace escaped newlines by actual newlines for readability
        sed -i.bak -E -e 's/\\n/\'$'\n/g' "$stdout_path"
        rm "$stdout_path.bak"
    else
```

leonardoalt commented:
> For me it outputs this locally:
> 
> ```json
> {"errors":[{"component":"general","errorCode":"3805","formattedMessage":"Warning: This is a pre-release compiler version, please do not use it in production.\n","message":"This is a pre-release compiler version, please do not use it in production.","severity":"warning","type":"Warning"}],"sources":{"A":{"id":0}}}
> ```

It's a command line test though, not a standard json. When I run via the cli, I also get the prerelase warning. But CI fails saying the message "Compiler run successful, no output requested."

leonardoalt commented:
Ok hopefully now all the trailing whitespace/counterexample mess is fixed.

cameel commented:
But it's the same for non-json tests. The `else` part removes the warning in that case but still does not remove that message from stderr. And it's on stderr only in release builds.

https://github.com/ethereum/solidity/blob/develop/test/cmdlineTests.sh#L110-L131

cameel commented:
You could change the script to strip it but I think that a simpler workaround would be just to add something to your test to cause an unrelated warnings. This way there's always some output.

leonardoalt commented:
@cameel I'm not sure we're talking about the same thing. The test fails on CI because of this expectation mismatch:
```
- model_checker_engine_none/
Incorrect output on stderr received. Expected:

But got:
Compiler run successful, no output requested.
```
It's not about the prerelease warning, it's about `no output requested`.

cameel commented:
We're talking about the same thing :) Whether `no output requested` is there or not depends on the presence of the pre-release warning.
- In CI it's probably a force release build so there's no warning and stderr contains `no output requested`. Test fails.
- Locally it's a pre-release build. Stderr contains the pre-release warning. But the script strips that warning and stderr becomes empty. Test succeeds.

leonardoalt commented:
> We're talking about the same thing :) Whether `no output requested` is there or not depends on the presence of the pre-release warning.

Ohhh ok, sorry, got it now.
Added something that yields a warning.

cameel commented on test/cmdlineTests/model_checker_engine_none/input.sol:
```diff
@@ -0,0 +1,10 @@
+contract test {
+	// `y` is here just to generate a warning, otherwise the command line test
+	// expectation fails with "Obtained no output requested".
+    function f(uint x, uint y) public pure {
```
You could also just remove the version pragma and license comment and get a warning that's clearly unrelated to the example.

leonardoalt replied:
Right, did that instead.

cameel reviewed (dismissed):
Looks like it's just a matter of getting that last test to pass now so I'm marking as approved.

leonardoalt commented:
@cameel https://app.circleci.com/pipelines/github/ethereum/solidity/9381/workflows/3cf1143b-4b45-4494-929c-fb5b7262b329/jobs/472705
the solution doesn't seem to work, unless I got it wrong. CI still complains about nothing on `stderr`, even though the `err` file is there.

cameel commented:
> even though the err file is there.

Is it? I don't see `model_checker_engine_none/err` in the diff.

leonardoalt commented:
> Is it? I don't see `model_checker_engine_none/err` in the diff.

Maybe I should stop for the week...

axic commented on Changelog.md:
```diff
@@ -12,6 +12,8 @@ Compiler Features:
  * SMTChecker: Support variables ``block``, ``msg`` and ``tx`` in the CHC engine.
  * Control Flow Graph: Print warning for non-empty functions with unnamed return parameters that are not assigned a value in all code paths.
+ * Command Line Interface: New option ``model-checker-engine`` allows to choose a specific SMTChecker engine. Options are ``all`` (default), ``bmc``, ``chc`` and ``none``.
+ * Standard JSON: New option ``settings.modelCheckerEngine`` allows to choose a specific SMTChecker engine. Options are ``all`` (default), ``bmc``, ``chc`` and ``none``.
```
We could consider defaulting this to `none` and removed the `pragma experimental SMTChecker;`. So then SMT is always enabled, but not ran by default as there's no engine selected.

cameel replied:
Would that work when the binary is compiled with Z3 and CVC4 turned off?

leonardoalt replied:
We could actually do that.
@cameel I guess that would work similar to now when you activate `pragma experimental SMTChecker;` but there's no solver.
The only difference to now would be that it would be enabled for all sources at the same time, whereas now it's enabled only for the sources with the pragma.
